ts/components/conversation/Quote.tsx

```diff
@@ -0,0 +1,169 @@
+import React from 'react';
+import classnames from 'classnames';
+
+// @ts-ignore
+import Mime from '../../../js/modules/types/mime';
+
+
+interface Props {
+  attachments: Array<QuotedAttachment>;
+  authorColor: string;
+  authorProfileName?: string;
+  authorTitle: string;
+  i18n: (key: string, values?: Array<string>) => string;
+  isIncoming: boolean;
+  openQuotedMessage?: () => void;
+  quoterAuthorColor?: string;
+  text: string;
+}
+
+interface QuotedAttachment {
+  contentType: string;
+  fileName: string;
+  /* Not included in protobuf */
+  isVoiceMessage: boolean;
+  thumbnail?: Attachment;
+}
+
+interface Attachment {
+  contentType: string;
+  /* Not included in protobuf, and is loaded asynchronously */
+  objectUrl?: string;
+}
+
+function validateQuote(quote: Props): boolean {
+  if (quote.text) {
+    return true;
+  }
+
+  if (quote.attachments && quote.attachments.length > 0) {
+    return true;
+  }
+
+  return false;
+}
+
+function getObjectUrl(thumbnail: Attachment | undefined): string | null {
+  if (thumbnail && thumbnail.objectUrl) {
+    return thumbnail.objectUrl;
+  }
+
+  return null;
+}
+
+export class Quote extends React.Component<Props, {}> {
+  public renderImage(url: string, icon?: string) {
+    const iconElement = icon
+      ? <div className={classnames('icon', icon)} />
+      : null;
+
+    return (
+      <div className="icon-container">
+        <div className="inner">
+          <img src={url} />
+          {iconElement}
+        </div>
+      </div>
+    );
+  }
+
+  public renderIcon(icon: string) {
+    const { authorColor, isIncoming, quoterAuthorColor } = this.props;
+
+    const backgroundColor = isIncoming ? 'white' : authorColor;
+    const iconColor = isIncoming ? quoterAuthorColor : 'white';
+
+    return (
+      <div className="icon-container">
+        <div className={classnames('circle-background', backgroundColor)} />
+        <div className={classnames('icon', icon, iconColor)} />
```

Comment:
Interesting, I’ve never used `classNames` like this. We’d always do:

```javascript
classNames({
  'circle-background': true,
  [icon]: icon,
  [iconColor]: iconColor,
});
```

This is clever. Will have to get used to it :wink: